File: Array/Sorting/Merge-Sorting/MergeSort.java
import java.util.Scanner;

public class MergeSort {
    public static void main(String[] args) {
        try (Scanner sc = new Scanner(System.in)) {
            System.out.print("Enter the size of the array: ");
            int n = sc.nextInt();

            int[] arr = new int[n];
            System.out.println("Enter the elements of the array:");
            for (int i = 0; i < n; i++) {
                arr[i] = sc.nextInt();
            }

            System.out.print("Original array: ");
            for (int num : arr) {
                System.out.print(num + " ");
            }
            System.out.println();

            int[] temp = new int[n];

            for (int size = 1; size < n; size *= 2) {
                for (int left = 0; left < n - size; left += 2 * size) {
                    int mid = left + size - 1;
                    int right = Math.min(left + 2 * size - 1, n - 1);

                    int i = left, j = mid + 1, k = left;

                    while (i <= mid && j <= right) {
                        if (arr[i] <= arr[j]) {
                            temp[k++] = arr[i++];
                        } else {
                            temp[k++] = arr[j++];
                        }
                    }

                    while (i <= mid) {
                        temp[k++] = arr[i++];
                    }

                    while (j <= right) {
                        temp[k++] = arr[j++];
                    }

                    for (int x = left; x <= right; x++) {
                        arr[x] = temp[x];
                    }
                }
            }

            System.out.print("Sorted array: ");
            for (int num : arr) {
                System.out.print(num + " ");
            }
            System.out.println();
        }
    }
}

// This code implements the merge sort algorithm in Java.

File: Array/Sorting/Merge-Sorting/MergeSort.py

# Input
n = int(input("Enter the size of the array: "))
arr = []
print("Enter the elements of the array:")
for _ in range(n):
    arr.append(int(input()))

print("Original array:", *arr)

# Merge Sort logic
temp = [0] * n
size = 1

while size < n:
    for left in range(0, n - size, 2 * size):
        mid = left + size - 1
        right = min(left + 2 * size - 1, n - 1)

        i, j, k = left, mid + 1, left

        while i <= mid and j <= right:
            if arr[i] <= arr[j]:
                temp[k] = arr[i]
                i += 1
            else:
                temp[k] = arr[j]
                j += 1
            k += 1

        while i <= mid:
            temp[k] = arr[i]
            i += 1
            k += 1

        while j <= right:
            temp[k] = arr[j]
            j += 1
            k += 1

        for x in range(left, right + 1):
            arr[x] = temp[x]

    size *= 2

print("Sorted array:", *arr)

# This code implements the merge sort algorithm in Python.
